# Exploring eBay Car Sales Data
In this project, we will be cleaning and analyzing 50,000 data points of used car sales from German eBay.

The main goal of this project is to familiarize ourselves with `Pandas`, `NumPy` and `Jupyter Notebook`.

### Data Dictionary
The data dictionary is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.* 

We begin by first importing the `Pandas` and `NumPy` libraries:

In [1]:
import pandas as pd
import numpy as np

We then read in our dataset using the `read_csv()` using the encoding `Latin-1`:

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

# Cleaning Column Names

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

We list the columns above, with their respective values and types. We can see that some columns contain `NULL` values, such as `vehicleType`, `gearbox`, `model`, `fuelType`, and `notRepairedDamage`. However, none of these have more than 20% `NULL` values. 

A majority of the columns contain `string` types, as represented by the type `object`. However, some of these values such as `price`, and `odometer` contain values that can be represented as `integer` types.

Also, the column names are in `camelCase`, which is not the Python standard of `snake_case`. 


First, we want to rename the columns using `snake_case`, as well as change some names to make them easier to work with:

In [4]:
print(autos.columns)
print('\n')

col_copy = autos.columns

col_copy = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

# Assign new cleaned columns to our dataset
autos.columns = col_copy

print(autos.columns)

# Display first 5 rows
autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Above, we converted our column names from `camelCase` to `snake_case`. We also changed some of the column names to make them easier to work with such as `yearOfRegistration` to `registration_year`, and `nrOfPictures` to `num_pictures`.

# Initial Exploration and Cleaning
Now, we will further explore our data to see what else should be cleaned. We will look for columns that contain mostly the same values. These will be dropped as they would hold no useful information. We are also looking for numeric data that is stored as text, which will need to be converted during cleaning.

In [5]:
autos.head(10)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   
5  privat    Angebot  $7,900     test          bus               2006   
6  privat    Angebot    $300     test    limousine               1995   
7  privat    Angebot  $1,990  control    limousine               1998   
8  privat    Angebot    $250     test          NaN               2000   
9  privat    Angebot    $590  control          bus               1997   

     gearbox  power_ps    model   odometer  registration_month fuel_type  \
0    manuell       158   andere  150,000km                   3       lpg   
1  automatik       286      7er  150,000km                   6    benzin   
2    manuell       102     golf   70,000km                   7    benzin   
3  automatik        71   fortwo   70,000km                   6    benzin   
4    manuell         0    focus  150,000km                   7    benzin   
5  automatik       150  voyager  150,000km                   4    diesel   
6    manuell        90     golf  150,000km                   8    benzin   
7    manuell        90     golf  150,000km                  12    diesel   
8    manuell         0    arosa  150,000km                  10       NaN   
9    manuell        90   megane  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50  
5        22962  2016-04-06 09:45:21  
6        31535  2016-03-23 02:48:59  
7        53474  2016-04-07 03:17:32  
8         7426  2016-03-26 18:18:10  
9        15749  2016-04-06 10:46:35

As we can see, the `seller`, `offer_type`, and `num_pictures` columns have the same values. We will use `autos.unique()` to get more information on these columns before making a decision:

In [6]:
# seller information
print(autos['seller'].unique())
print('These translate to \"private" and \"commerical".')
print('\n')

# offer_type information
print(autos['offer_type'].unique())
print('These translate to \"offer" and \"request".')
print('\n')

# num_pictures information
print(autos['num_pictures'].unique())
print('There are no unique values besides 0, we will drop this column, as it doesn\'t provide any useful information.')

autos = autos.drop('num_pictures', axis = 1)

autos.head()

['privat' 'gewerblich']
These translate to "private" and "commerical".


['Angebot' 'Gesuch']
These translate to "offer" and "request".


[0]
There are no unique values besides 0, we will drop this column, as it doesn't provide any useful information.


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00        79588   
1         bmw              nein  2016-04-04 00:00:00        71034   
2  volkswagen              nein  2016-03-26 00:00:00        35394   
3       smart              nein  2016-03-12 00:00:00        33729   
4        ford              nein  2016-04-01 00:00:00        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

From the unique data above, we can see that the `seller` and `offer_type` columns contain relevant information. However, the `num_pictures` column contains nothing but 0's, so it was dropped. 

We can also see from the table above that the `price` and `odometer` columns need to be cleaned in order to be `integer` types. We do this by removing the non-numeric characters and then converting them to numeric types.

In [7]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(float)

autos['odometer'] = autos['odometer'].str.replace(',', '').str.replace('km', '').astype(int)

autos.rename({'price':'price_usd'}, axis = 1, inplace = True)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_usd  ab_test vehicle_type  registration_year  \
0  privat    Angebot     5000.0  control          bus               2004   
1  privat    Angebot     8500.0  control    limousine               1997   
2  privat    Angebot     8990.0     test    limousine               2009   
3  privat    Angebot     4350.0  control   kleinwagen               2007   
4  privat    Angebot     1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00        79588   
1         bmw              nein  2016-04-04 00:00:00        71034   
2  volkswagen              nein  2016-03-26 00:00:00        35394   
3       smart              nein  2016-03-12 00:00:00        33729   
4        ford              nein  2016-04-01 00:00:00        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Above, we removed the non-numeric characters from the values in both the `price` and `odometer` columns. We also renamed the columns so that we will not be confused as to what the values are, such as the `price` in USD (`price_usd`), and the `odometer` in km (`odometer_km`).

# Exploring the Odometer and Price Columns
We will now further explore the `odometer_km` and `price_usd` columns for any irregularities. We will be using different methods such as `describe()`, `shape`, `value_counts()`, and even some `boolean indexing` to analyze the data further.

### Odometer Analysis

In [8]:
# Unique values for odometer
print(autos['odometer_km'].unique().shape)
print('\n')

# Measures of central tendency for odometer
print(autos['odometer_km'].describe())
print('\n')

# Values
print(autos['odometer_km'].value_counts())

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


Every `odometer` value is rounded, which may be a result of pre-set values on the website. Also, we can see that there are more high-mileage cars for sale than low-mileage ones.

### Price Analysis

In [9]:
# Unique values for price
print(autos['price_usd'].unique().shape)
print('\n')

# Measures of central tendency for price
print(autos['price_usd'].describe())
print('\n')

# Values sorted descending
print('Prices')
print(autos['price_usd'].value_counts().sort_index(ascending = False))

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64


Prices
99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price_usd, Length: 2357, dtype: int64


From the measures of central tendency, we can see that there is a max of `$100,000,000`. When we inspect further, we see some values that are in the `$10,000,000` range. We can also see that `1,421` cars are being sold at `$0`, which could indicate an error in the data. Since eBay is a bidding website, we will be leaving values such as `$1` and similar, as bids can start at that value.

We will inspect the `price_usd` column further:

In [10]:
print(autos['price_usd'].value_counts().sort_index(ascending = False).head(20))

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64


We can see that there is a steady increase in `price_usd` up until `$350,000`, and then the price jumps to `$999,990`. Therefore, our cutoff point is `$350,000`.

In [11]:
autos = autos[autos['price_usd'].between(1, 351000)]
autos['price_usd'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

Now, our new distribution of `price_usd` includes values ranging from `$1` to `$350,000`.

# Exploring the Date Columns
There are five columns that represent dates. Some of them were added by the crawler, and some were from the website originally:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The `date_crawled`, `last_seen`, and `ad_created` columns are all being represented as string values by `Pandas`. 

In [12]:
autos[['date_crawled', 'last_seen', 'ad_created']][:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

We can see that the dates are represented by the `YYYY-MM-DD` format. We will analyze them further below:

In [13]:
# date_crawled
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

We use `.sort_index()` to sort our dates. We can see that the data was crawled daily from about `March to April 2016`. 

In [14]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The `last_seen` column shows when the crawler last saw this car. This means that if it wasn't seen again, the car was either sold or the listing was taken down. We can see that on the `5th of April`, the value spiked `5x` from the `4th of April`. This can either mean that there was a spike in sales on that day, or more likely, the crawler was finished and spiked the values itself.

In [15]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

From this column, we can see that some of the listings were created in `2015`, with the oldest listing being created on `June 11, 2015`. 

Next, we will analyze the `registration_year` column.

# Dealing with Incorrect Registration Year Data
We will now analyze the `registration_year` column for any irregularities.

In [16]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can see that the minimum value in the `registration_year` column is `1000`, and the maximum value is `9999`. These are clearly irregularities, as no car could be registered in the year `1000`, or `9999`. Therefore, we will investigate further:

In [17]:
# First 20 values ascending
print(autos['registration_year'].sort_values().head(20))
print('\n')

# Last 20 values
print(autos['registration_year'].sort_values().tail(20))

22316    1000
49283    1001
24511    1111
32585    1800
10556    1800
22659    1910
45157    1910
30781    1910
28693    1910
3679     1910
21416    1927
22101    1929
11246    1931
2221     1934
2573     1934
39725    1937
21421    1937
23804    1937
26607    1937
26103    1938
Name: registration_year, dtype: int64


44772    2018
17928    2018
32040    2018
5763     2019
49185    2019
27578    2800
4549     4100
453      4500
42079    4800
4164     5000
22799    5000
49153    5000
24519    5000
27618    5911
8360     6200
25003    8888
49910    9000
8012     9999
38076    9999
33950    9999
Name: registration_year, dtype: int64


The data was crawled in `2016`. The highest possible registration date could be `2016`, or `2017` for new models. The lowest reasonable date was in `1910`. 

With this information, our range of values will be between `1910 and 2017`. Let's calculate the percentage of years outside of this range.

In [18]:
((~autos['registration_year'].between(1910, 2017)).sum() / autos.shape[0]) * 100

1.0130752599608772

Since `1%` of the `registration_year` values fall outside of our range, we can safely remove them from our data altogether.

In [19]:
autos = autos[autos['registration_year'].between(1910, 2017)]
autos['registration_year'].value_counts().sort_index()

1910       5
1927       1
1929       1
1931       1
1934       2
        ... 
2013     803
2014     663
2015     392
2016    1220
2017    1392
Name: registration_year, Length: 79, dtype: int64

We can now see that our lowest value is `1910` and our highest value is `2017`. 

# Exploring Price by Brand
Next, we will analyze the average `price` in terms of `brand`.

We will be using the `brands` that have a presence of `5%` or greater.

In [20]:
print(autos['brand'].value_counts(normalize = True))

percentages = autos['brand'].value_counts(normalize = True)
                    
common_brands = percentages[percentages > 0.05].index

common_brands

volkswagen        0.212593
bmw               0.108814
opel              0.108377
mercedes_benz     0.095896
audi              0.086098
ford              0.069831
renault           0.047677
peugeot           0.029601
fiat              0.025836
seat              0.018618
skoda             0.016163
nissan            0.015310
mazda             0.015248
smart             0.014312
citroen           0.014104
toyota            0.012668
hyundai           0.009985
sonstige_autos    0.009631
volvo             0.009090
mini              0.008695
mitsubishi        0.008154
honda             0.007905
kia               0.007114
alfa_romeo        0.006636
porsche           0.005949
suzuki            0.005866
chevrolet         0.005637
chrysler          0.003515
dacia             0.002663
daihatsu          0.002496
jeep              0.002226
subaru            0.002101
land_rover        0.002059
saab              0.001623
daewoo            0.001560
jaguar            0.001519
trabant           0.001394
r

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

The brands we will be using include `volkswagen`, `bmw`, `opel`, `mercedes_benz`, `audi`, and `ford`. 

We can see that the most popular brand is `volkswagen`, having a value of `21%`. 

Now, we will aggregate this data, finding the average `price` per brand, and creating a dictionary with these values.

In [21]:
avg_price_to_brand = {}

for brand in common_brands:
    brand_bool = autos[autos['brand'] == brand]
    mean_price = brand_bool['price_usd'].mean()
    
    avg_price_to_brand[brand] = round(float(mean_price), 2)
    
avg_price_to_brand

{'volkswagen': 5351.4,
 'bmw': 8284.05,
 'opel': 2952.41,
 'mercedes_benz': 8528.27,
 'audi': 9239.32,
 'ford': 3732.27}

# Storing Aggregate Data in a DataFrame
We now have the top `brands` and their average `prices`. We can see that the `audi`, `mercedes_benz`, and `bmw` brands have the highest average price. The lowest average price belongs to `ford`.

Next, we will compare these average prices with the mileage on the brands to see if brand and mileage have and effect on price.

In [22]:
# We first calculate the average mileage to brand as we did price to brand
avg_mileage_to_brand = {}

for brand in common_brands:
    brand_bool = autos[autos['brand'] == brand]
    mean_mileage = brand_bool['odometer_km'].mean()
    
    avg_mileage_to_brand[brand] = round(float(mean_mileage), 2)
    
avg_mileage_to_brand

{'volkswagen': 128928.08,
 'bmw': 132666.79,
 'opel': 129415.55,
 'mercedes_benz': 130962.04,
 'audi': 129406.86,
 'ford': 124255.29}

Now that we have the average mileage per brand, we will compare these to the average prices.

In [23]:
# Convert both dictionaries to a series
avg_price_series = pd.Series(avg_price_to_brand)
avg_mileage_series = pd.Series(avg_mileage_to_brand)

# Create a dataframe for brand and price and assign mileage as a column
price_to_mileage_dataframe = pd.DataFrame(avg_price_series, columns = ['mean_price_usd'])
price_to_mileage_dataframe['mean_mileage_km'] = avg_mileage_series

price_to_mileage_dataframe

mean_price_usd  mean_mileage_km
volkswagen            5351.40        128928.08
bmw                   8284.05        132666.79
opel                  2952.41        129415.55
mercedes_benz         8528.27        130962.04
audi                  9239.32        129406.86
ford                  3732.27        124255.29

# Conclusion
Now that we have the top 6 brands, their average prices, and their average mileages, we can make inferences on our data. 

We can see that the brands all have average mileages that differ by a few thousand miles, with `ford` differing some more. However, we can infer that the `brand` of a car does have an effect on the `price`. 

`Audi` has an average price of `$9,239.32` and an average mileage of about `129,000km`. `Opel`, on the other hand, has an average price of `$2,952` and an average mileage of about `129,000km` as well. `Audi` is obviously seen as a more luxurious brand, therefore leaving it with a higher average price than `Opel`. 

We can infer that `price` is moreso influenced by the `brand` instead of by the mileage.